cmd/xl-storage.go

```diff
@@ -343,31 +340,53 @@ func isQuitting(endCh chan struct{}) bool {
 	}
 }
 
-func (s *posix) waitForLowActiveIO() {
+func (s *xlStorage) waitForLowActiveIO() {
 	for atomic.LoadInt32(&s.activeIOCount) >= s.maxActiveIOCount {
 		time.Sleep(lowActiveIOWaitTick)
 	}
 }
 
-func (s *posix) CrawlAndGetDataUsage(ctx context.Context, cache dataUsageCache) (dataUsageCache, error) {
+func (s *xlStorage) CrawlAndGetDataUsage(ctx context.Context, cache dataUsageCache) (dataUsageCache, error) {
 	dataUsageInfo, err := updateUsage(ctx, s.diskPath, cache, s.waitForLowActiveIO, func(item Item) (int64, error) {
-		// Look for `xl.json' at the leaf.
-		if !strings.HasSuffix(item.Path, SlashSeparator+xlMetaJSONFile) {
-			// if no xl.json found, skip the file.
+		// Look for `xl.meta/xl.json' at the leaf.
+		if !strings.HasSuffix(item.Path, SlashSeparator+xlStorageFormatFile) &&
+			!strings.HasSuffix(item.Path, SlashSeparator+xlStorageFormatFileV1) {
+			// if no xl.meta/xl.json found, skip the file.
 			return 0, errSkipFile
 		}
 
-		xlMetaBuf, err := ioutil.ReadFile(item.Path)
+		buf, err := ioutil.ReadFile(item.Path)
 		if err != nil {
 			return 0, errSkipFile
 		}
 
-		meta, err := xlMetaV1UnmarshalJSON(ctx, xlMetaBuf)
+		volume, path := path2BucketObject(strings.TrimSuffix(strings.TrimPrefix(item.Path, s.diskPath),
+			SlashSeparator+xlStorageFormatFile))
+		if isXLVersion2(buf) {
+			var xlMeta XLMetaV2
+			if _, err = xlMeta.UnmarshalMsg(buf[4:]); err != nil {
+				return 0, errSkipFile
+			}
+			fis, err := xlMeta.ListFileInfos(volume, path)
```

Comment:
This returns an error if there is a delete marker in there. Let's make it simpler:
```suggestion
			return xlMeta.TotalSize(), nil
```
Add:

```Go
// TotalSize returns the total size of all versions.
func (z XLMetaV2) TotalSize() int64 {
	var total int64
	for i := range z.ObjectJournals {
		obj := z.ObjectJournals[i].Object
		if obj == nil {
			continue
		}
		total += obj.StatSize
	}
	return total
}
```